packages/tailwindcss/src/index.ts

```diff
@@ -306,25 +306,20 @@ async function parseCss(
 
   let designSystem = buildDesignSystem(theme)
 
-  let configs = await Promise.all(
-    configPaths.map(async (configPath) => ({
-      path: configPath,
-      config: await loadConfig(configPath),
-    })),
-  )
-
-  let plugins = await Promise.all(
-    pluginPaths.map(async ([pluginPath, pluginOptions]) => ({
-      path: pluginPath,
-      plugin: await loadPlugin(pluginPath),
-      options: pluginOptions,
-    })),
+  // Apply hooks from backwards compatibility layer. This function takes a lot
+  // of random arguments because it really just needs access to "the world" to
+  // do whatever ungodly things it needs to do to make things backwards
+  // compatible without polluting core.
+  await applyCompatibilityHooks(
+    designSystem,
+    ast,
+    pluginPaths,
+    loadPlugin,
+    configPaths,
+    loadConfig,
+    globs,
```

Comment:
Some thoughts:

1. This should definitely take an object as a param
2. I think this object should (eventually) have a name of some kind. Especially given that it basically represents structured data / knowledge about the input CSS. Not immediately but maybe a `Tailwind` object or something that metadata attached to it about the current CSS? I think this goes to the larger API design stuff we were talking about at one point. (idk could even call it `CSSFile` or something but that doesn't "feel" quite right)
3. I think we should do the AST walk and validation stuff for `@plugin` and `@config` in this function too (maybe). It already takes `ast` as an argument and you could eliminate `pluginPaths` and `configPaths` options. The downside here being there's an extra AST walk but we're already doing multiple walks. If we're able to optimize those maybe on extra walk won't be as big of a deal.

Reply:
Yeah moving the `@plugin` and `@config` extraction into the compat hooks would be nice since currently that information is only used in that later.

Later, when there is a CSS `@plugin` API, we could leave `@plugin 'path'` untouched in the core implementation and can then separate between the core v4 implementation and the compat one more cleanly as well.